### PR TITLE
Refactor ssd cache configs

### PR DESCRIPTION
**Pull request by zacw7**

### CODE REVIEW

File: velox/common/caching/SsdCache.cpp [hunk 1]
```diff
@@ -33,14 +33,12 @@ SsdCache::SsdCache(
     std::string_view filePrefix,
     uint64_t maxBytes,
     int32_t numShards,
-    folly::Executor* executor,
-    int64_t checkpointIntervalBytes,
-    bool disableFileCow,
-    bool checksumEnabled,
-    bool checksumReadVerificationEnabled)
+    Options options,
```
Comment (amitkdutta) [inline velox/common/caching/SsdCache.cpp:36]: Nice. This is much cleaner.

**Review by amitkdutta — APPROVED** [review]
Thanks @zacw7. Looks great.

File: velox/common/caching/SsdFile.h [hunk 2]
```diff
@@ -27,6 +27,21 @@ DECLARE_bool(ssd_verify_write);
 
 namespace facebook::velox::cache {
 
+struct Options {
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.h:30]: Let's have SSDFile::Config? Thanks!

File: velox/common/caching/SsdCache.h [hunk 3]
```diff
@@ -49,6 +49,13 @@ class SsdCache {
       bool checksumWriteEnabled = false,
       bool checksumReadVerificationEnabled = false);
 
+  SsdCache(
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdCache.h:52]: Let's have SsdCache::Config which include all the parameters? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@zacw7 thanks % comments.

**Review by xiaoxmeng — APPROVED** [review]
@zacw7 LGTM % nits. Thanks!

File: velox/common/caching/SsdCache.h [hunk 4]
```diff
@@ -49,6 +102,8 @@ class SsdCache {
       bool checksumWriteEnabled = false,
       bool checksumReadVerificationEnabled = false);
 
+  SsdCache(Config& config);
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdCache.h:105]:
```
SsdCache(const Config& config);
```

File: velox/common/caching/SsdFile.h [hunk 5]
```diff
@@ -230,19 +230,57 @@ struct SsdCacheStats {
...
-  SsdFile(
-      const std::string& filename,
-      int32_t shardId,
-      int32_t maxRegions,
-      int64_t checkpointInternalBytes = 0,
-      bool disableFileCow = false,
-      bool checksumEnabled = false,
-      bool checksumReadVerificationEnabled = false,
-      folly::Executor* executor = nullptr);
+  SsdFile(Config& config);
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdFile.h:283]:
```
SsdFile(const Config& config);
```

File: velox/common/caching/SsdCache.cpp [hunk 6]
```diff
@@ -38,46 +38,61 @@ SsdCache::SsdCache(
     bool disableFileCow,
     bool checksumEnabled,
     bool checksumReadVerificationEnabled)
-    : filePrefix_(filePrefix),
-      numShards_(numShards),
-      groupStats_(std::make_unique<FileGroupStats>()),
-      executor_(executor) {
+    : groupStats_(std::make_unique<FileGroupStats>()) {
```
[the remark this answers is not in the document]
Comment (zacw7) [reply velox/common/caching/SsdCache.cpp:41]: Yes. But `SsdCache(Config& config)` is expecting a lvalue which is incompatiable with this approach. I think what we can do is to simply the config_ creation to the initialization list.

File: velox/common/caching/SsdFile.h [hunk 7]
```diff
@@ -448,26 +488,11 @@ class SsdFile {
...
-  // If true, checksum read verification from SSD is enabled.
-  const bool checksumReadVerificationEnabled_;
-
   // Serializes access to all private data members.
   mutable std::shared_mutex mutex_;
 
-  // Shard index within 'cache_'.
-  int32_t shardId_;
+  // SSD cache file config.
+  Config config_;
```
[the remark this answers is not in the document]
Comment (zacw7) [reply velox/common/caching/SsdFile.h:495]: If error happens while reading checkpoint, `config_.checkpointIntervalBytes` needs to be set to 0 to disable it.

File: velox/benchmarks/tpch/TpchBenchmark.cpp [hunk 8]
```diff
@@ -233,12 +233,13 @@ class TpchBenchmark {
         constexpr int32_t kNumSsdShards = 16;
         cacheExecutor_ =
             std::make_unique<folly::IOThreadPoolExecutor>(kNumSsdShards);
-        ssdCache = std::make_unique<cache::SsdCache>(
+        cache::SsdCache::Config config(
```
[the remark this answers is not in the document]
Comment (zacw7) [reply velox/benchmarks/tpch/TpchBenchmark.cpp:236]: `config_.checksumReadVerificationEnabled` might need to be set to false if checksumEnabled is false.

File: velox/common/caching/SsdCache.cpp [hunk 9]
```diff
@@ -24,60 +24,85 @@
...
 #include <numeric>
+#include <utility>
 
 using facebook::velox::common::testutil::TestValue;
 
 namespace facebook::velox::cache {
 
+SsdFile::Config toSsdFileConfig(
+    const SsdCache::Config& config,
+    const int32_t fileMaxRegions,
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdCache.cpp:35]:
```
s/const int32_t fileMaxRegions/int32_t fileMaxRegions/
s/const int32_t shardId/int32_t shardId/
```
Do not need const qualifier to scalar function input args

File: velox/common/caching/SsdCache.cpp [hunk 10]
```diff
@@ -24,60 +24,85 @@
...
-    int64_t checkpointIntervalBytes,
+    uint64_t checkpointIntervalBytes,
     bool disableFileCow,
     bool checksumEnabled,
     bool checksumReadVerificationEnabled)
-    : filePrefix_(filePrefix),
-      numShards_(numShards),
-      groupStats_(std::make_unique<FileGroupStats>()),
-      executor_(executor) {
+    : groupStats_(std::make_unique<FileGroupStats>()),
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdCache.cpp:57]: You can call
```
SsdCache::SsdCache() : SsdCache({std::string(filePrefix),
           maxBytes,
           numShards,
           executor,
           checkpointIntervalBytes,
           disableFileCow,
           checksumEnabled,
           checksumReadVerificationEnabled})
```

File: velox/common/caching/SsdCache.h [hunk 11]
```diff
@@ -119,11 +174,11 @@ class SsdCache {
...
   }
 
-  const std::string filePrefix_;
-  const int32_t numShards_;
+  void initialize();
+
   // Stats for selecting entries to save from AsyncDataCache.
   const std::unique_ptr<FileGroupStats> groupStats_;
-  folly::Executor* const executor_;
+  Config config_;
```
Comment (xiaoxmeng) [inline velox/common/caching/SsdCache.h:181]:
```
const Config config_;?
```

**Review by xiaoxmeng — APPROVED** [review]
@zacw7 thanks!

**Pull request closed without merging**